File: imtop/core/metrics.py
"""Agreement and geometry metrics between the manual (GT) and automatic masks.

The key names are a **published schema**: ``results_all.csv`` and the Monte
Carlo scripts read them, so they never change. The ``*_mm`` / ``*_mm2`` suffix
is historical, when no scale is set those values are in pixels. Callers get
the truth from :func:`metrics_info` and must label the numbers accordingly.
"""
from __future__ import annotations

import numpy as np
from scipy.spatial.distance import cdist

from .geometry import axis_aligned_bbox, polygon_perimeter

# Canonical key order, shared by the UI, the JSON export and the batch CSV.
METRIC_KEYS = [
    "dice", "iou", "area_gt_mm2", "area_auto_mm2", "area_rel_err",
    "cri_lambda3", "under_cov_mm2", "over_cov_mm2", "hausdorff_mm",
    "avg_surf_dist_mm", "perimeter_gt_mm", "perimeter_auto_mm",
    "compactness_gt", "compactness_auto", "bbox_gt_major_mm",
    "bbox_gt_minor_mm", "bbox_auto_major_mm", "bbox_auto_minor_mm",
    "aspect_ratio_gt", "aspect_ratio_auto", "bbox_auto_x_mm", "bbox_auto_y_mm",
]

# Which keys carry a length or an area (and therefore change unit with the
# scale). Everything else is dimensionless.
LENGTH_KEYS = frozenset(k for k in METRIC_KEYS if k.endswith("_mm"))
AREA_KEYS = frozenset(k for k in METRIC_KEYS if k.endswith("_mm2"))

# Coverage cost index weight: under-coverage (wound left exposed) counts three
# times an equal area of over-coverage (excess on healthy skin).
CRI_LAMBDA = 3


def metrics_info(px_per_mm: float | None) -> dict:
    """How to read the numbers: calibrated or not, and the unit labels."""
    calibrated = bool(px_per_mm)
    return {
        "calibrated": calibrated,
        "px_per_mm": float(px_per_mm) if calibrated else None,
        "length_unit": "mm" if calibrated else "px",
        "area_unit": "mm²" if calibrated else "px²",
    }


def compute_metrics(gt_mask: np.ndarray | None,
                    auto_mask: np.ndarray | None,
                    gt_contour: np.ndarray | None,
                    auto_contour: np.ndarray | None,
                    px_per_mm: float | None) -> dict:
    """All metrics for one GT/auto pair. Empty dict when either mask is missing.

    Keys starting with ``_`` are internal helpers (pixel bounding boxes) and are
    stripped by :func:`public_metrics` before anything leaves the process.
    """
    gm, sm, gc, sc = gt_mask, auto_mask, gt_contour, auto_contour
    if gm is None or sm is None:
        return {}
    s = px_per_mm or 1.0
    m: dict = {}
    inter = np.logical_and(gm, sm).sum()
    union = np.logical_or(gm, sm).sum()
    m["dice"] = round(2 * inter / (gm.sum() + sm.sum() + 1e-9), 4)
    m["iou"] = round(inter / (union + 1e-9), 4)
    m["area_gt_mm2"] = round(gm.sum() / s ** 2, 2)
    m["area_auto_mm2"] = round(sm.sum() / s ** 2, 2)
    m["area_rel_err"] = round(abs(m["area_gt_mm2"] - m["area_auto_mm2"])
                              / (m["area_gt_mm2"] + 1e-9), 4)
    # Asymmetric coverage. W = GT (the wound), P = auto (the patch).
    under_px = int(np.logical_and(gm, ~sm).sum())   # wound left uncovered
    over_px = int(np.logical_and(sm, ~gm).sum())    # excess on healthy skin
    area_gt_px = int(gm.sum())
    m["cri_lambda3"] = round((CRI_LAMBDA * under_px + over_px) / (area_gt_px + 1e-9), 4)
    if px_per_mm:
        m["under_cov_mm2"] = round(under_px / px_per_mm ** 2, 2)
        m["over_cov_mm2"] = round(over_px / px_per_mm ** 2, 2)
    else:
        m["under_cov_mm2"] = None
        m["over_cov_mm2"] = None
    if gc is not None and sc is not None:
        D = cdist(gc, sc)
        m["hausdorff_mm"] = round(max(D.min(1).max(), D.min(0).max()) / s, 3)
        m["avg_surf_dist_mm"] = round(0.5 * (D.min(1).mean() + D.min(0).mean()) / s, 3)
        Pg = polygon_perimeter(gc)
        Pa = polygon_perimeter(sc)
        m["perimeter_gt_mm"] = round(Pg / s, 2)
        m["perimeter_auto_mm"] = round(Pa / s, 2)
        m["compactness_gt"] = round(4 * np.pi * gm.sum() / (Pg ** 2 + 1e-9), 4)
        m["compactness_auto"] = round(4 * np.pi * sm.sum() / (Pa ** 2 + 1e-9), 4)
        gx0, gy0, gx1, gy1 = axis_aligned_bbox(gc)
        ax0, ay0, ax1, ay1 = axis_aligned_bbox(sc)
        m["bbox_gt_major_mm"] = round(max(gx1 - gx0, gy1 - gy0) / s, 2)
        m["bbox_gt_minor_mm"] = round(min(gx1 - gx0, gy1 - gy0) / s, 2)
        m["bbox_auto_major_mm"] = round(max(ax1 - ax0, ay1 - ay0) / s, 2)
        m["bbox_auto_minor_mm"] = round(min(ax1 - ax0, ay1 - ay0) / s, 2)
        m["aspect_ratio_gt"] = round(min(gx1 - gx0, gy1 - gy0)
                                     / (max(gx1 - gx0, gy1 - gy0) + 1e-9), 4)
        m["aspect_ratio_auto"] = round(min(ax1 - ax0, ay1 - ay0)
                                       / (max(ax1 - ax0, ay1 - ay0) + 1e-9), 4)
        m["bbox_auto_x_mm"] = round((ax1 - ax0) / s, 2)   # A (X): auto bbox width
        m["bbox_auto_y_mm"] = round((ay1 - ay0) / s, 2)   # B (Y): auto bbox height
        m["_bbox_gt"] = (gx0, gy0, gx1, gy1)
        m["_bbox_auto"] = (ax0, ay0, ax1, ay1)
    return m


def public_metrics(m: dict) -> dict:
    """Drop the internal ``_``-prefixed helpers."""
    return {k: v for k, v in m.items() if not k.startswith("_")}


def coverage_extras(gt_mask: np.ndarray, auto_mask: np.ndarray, metrics: dict) -> dict:
    """Scale-free coverage columns used by the batch CSV (``EXTRA_KEYS``)."""
    area_gt_px = int(gt_mask.sum())
    under_px = int(np.logical_and(gt_mask, ~auto_mask).sum())
    over_px = int(np.logical_and(auto_mask, ~gt_mask).sum())
    out = {
        "n_gt_px": area_gt_px,
        "n_auto_px": int(auto_mask.sum()),
        "under_px": under_px,
        "over_px": over_px,
        "under_cov_frac": round(under_px / (area_gt_px + 1e-9), 6),
        "over_cov_frac": round(over_px / (area_gt_px + 1e-9), 6),
    }
    if "area_rel_err" in metrics:
        out["patch_area_err_pct"] = round(100 * metrics["area_rel_err"], 4)
    return out


EXTRA_KEYS = ["n_gt_px", "n_auto_px", "under_px", "over_px",
              "under_cov_frac", "over_cov_frac", "patch_area_err_pct"]

File: imtop/ui/js/metrics.js
/* ============================================================================
   metrics.js: step 5. The numbers the app exists to produce.

   The key strings are a published schema (imtop/core/metrics.py, results_all.csv
   and the Monte Carlo scripts read them): never rename one here. Grouping and
   wording are ours; the colour of a group is the colour its overlay uses.

   Each item says what it measures (`u`: 'len', 'area', or nothing for a ratio)
   and how many decimals it deserves (`p`). The unit itself is not written
   here: the "_mm" suffix in the keys is historical, and on an uncalibrated
   image the same numbers are pixels. getMetricsInfo() says which, once per
   run, and the labels are built from that answer.
   ========================================================================== */

const MGRPS = [
  { t: 'Overlap', g: 'ov', items: [
    { k: 'dice', l: 'Dice', p: 4 },
    { k: 'iou', l: 'IoU / Jaccard', p: 4 },
  ] },
  { t: 'Area', g: 'ar', items: [
    { k: 'area_gt_mm2', l: 'Manual area', u: 'area', p: 1 },
    { k: 'area_auto_mm2', l: 'Auto area', u: 'area', p: 1 },
    { k: 'area_rel_err', l: 'Rel. area error', p: 3 },
    { k: 'under_cov_mm2', l: 'Uncovered', u: 'area', p: 1, tip: 'Wound area the patch leaves uncovered (manual minus auto)' },
    { k: 'over_cov_mm2', l: 'Excess', u: 'area', p: 1, tip: 'Patch area on healthy skin (auto minus manual)' },
    { k: 'cri_lambda3', l: 'Cost index λ=3', p: 3, tip: 'Coverage cost: uncovered wound weighs 3 times the excess, relative to the wound area' },
  ] },
  { t: 'Distance', g: 'di', items: [
    { k: 'hausdorff_mm', l: 'Hausdorff', u: 'len', p: 2 },
    { k: 'avg_surf_dist_mm', l: 'Avg surface dist', u: 'len', p: 2 },
  ] },
  { t: 'Manual geometry', g: 'gm', items: [
    { k: 'perimeter_gt_mm', l: 'Perimeter', u: 'len', p: 2 },
    { k: 'compactness_gt', l: 'Compactness', p: 3 },
    { k: 'bbox_gt_major_mm', l: 'BBox major', u: 'len', p: 2 },
    { k: 'bbox_gt_minor_mm', l: 'BBox minor', u: 'len', p: 2 },
    { k: 'aspect_ratio_gt', l: 'Aspect ratio', p: 3 },
  ] },
  { t: 'Auto geometry', g: 'ga', items: [
    { k: 'perimeter_auto_mm', l: 'Perimeter', u: 'len', p: 2 },
    { k: 'compactness_auto', l: 'Compactness', p: 3 },
    { k: 'bbox_auto_major_mm', l: 'BBox major', u: 'len', p: 2 },
    { k: 'bbox_auto_minor_mm', l: 'BBox minor', u: 'len', p: 2 },
    { k: 'aspect_ratio_auto', l: 'Aspect ratio', p: 3 },
    { k: 'bbox_auto_x_mm', l: 'A (X)', u: 'len', p: 2 },
    { k: 'bbox_auto_y_mm', l: 'B (Y)', u: 'len', p: 2 },
  ] },
];

/* The unit an item is in right now. metInfo is the backend's word; until it
   has answered, the scale the contours came with is the next best thing. */
function metUnit(it) {
  if (!it.u) return '';
  if (metInfo) return it.u === 'area' ? metInfo.area_unit : metInfo.length_unit;
  const cal = !!pxPerMm;
  return it.u === 'area' ? (cal ? 'mm²' : 'px²') : (cal ? 'mm' : 'px');
}

function metLabel(it) {
  const u = metUnit(it);
  return u ? it.l + ' (' + u + ')' : it.l;
}

/* Pixels are whole numbers: a pixel area with a decimal is noise, a pixel
   length gets one. Everything else keeps the item's own precision. */
function fmtMetric(it, v) {
  if (v === undefined || v === null) return '-';
  if (typeof v !== 'number') return String(v);
  let p = it.p;
  if (it.u && metUnit(it).indexOf('px') === 0) p = it.u === 'area' ? 0 : 1;
  return v.toFixed(p);
}

function buildMetPanel() {
  const panel = document.getElementById('metPanel');
  panel.innerHTML = '';
  MGRPS.forEach(function (grp) {
    const color = MPAL[grp.g].m;
    const head = document.createElement('div');
    head.className = 'mg';
    head.style.color = color;
    head.textContent = grp.t;
    panel.appendChild(head);

    grp.items.forEach(function (it) {
      const row = document.createElement('div');
      row.className = 'mrow';
      row.style.setProperty('--gc', color);
      if (it.tip) row.title = it.tip;
      const label = document.createElement('span');
      label.className = 'ml';
      label.textContent = metLabel(it);
      const value = document.createElement('span');
      value.className = 'mv';
      value.style.color = color;
      value.textContent = fmtMetric(it, metData[it.k]);
      row.appendChild(label);
      row.appendChild(value);
      row.onclick = function () { clickMet(it.k, row); };
      panel.appendChild(row);
    });
  });
}

/* Clicking a row draws its overlay; clicking it again goes back to the plain
   two-contour view. */
function clickMet(key, row) {
  if (!metCnt.auto) { toast('Run the segmentation first'); return; }
  const rows = document.querySelectorAll('.mrow');
  for (let i = 0; i < rows.length; i++) rows[i].classList.remove('active');
  if (metKey === key) { metKey = null; dMet(); return; }
  metKey = key;
  row.classList.add('active');
  redraw();
}

/* Plain view: both contours, filled faintly. */
function dMet() {
  if (!iUrl) return;
  const f = beginFrame(), ctx = f.ctx;
  if (metCnt.gt) {
    const P = _toC(metCnt.gt, f.sc, f.ox, f.oy);
    _fillPoly(ctx, P, _rgba(COL['c-gt'], .09));
    strokeContour(ctx, metCnt.gt, f, COL['c-gt'], null);
  }
  if (metCnt.auto) {
    const P = _toC(metCnt.auto, f.sc, f.ox, f.oy);
    _fillPoly(ctx, P, _rgba(COL['c-auto'], .09));
    strokeContour(ctx, metCnt.auto, f, COL['c-auto'], [5, 3]);
  }
}

/* Overlay view: the geometry behind the selected number. */
function dMetVec() {
  if (!iUrl) return;
  const f = beginFrame(), ctx = f.ctx;
  const G = metCnt.gt ? _toC(metCnt.gt, f.sc, f.ox, f.oy) : null;
  const A = metCnt.auto ? _toC(metCnt.auto, f.sc, f.ox, f.oy) : null;
  drawMetricOverlay(ctx, metKey, G, A);
}

/* ── exports ──────────────────────────────────────────────────────────────── */
function saveRes() {
  be('saveResults', undefined, function (msg) {
    toast(msg || 'Saved ✓');
    if (msg && msg.indexOf('Saved') === 0) { resSaved = true; paintTabs(); }
  });
}

/* The report: a PDF, printed by the same engine that renders this page, so
   it needs nothing installed and no connection. Printing is asynchronous, the
   outcome arrives later on reportDone (an .html name typed in the dialog is
   written and answered at once instead). The step-6 sliders travel with the
   request because only this page knows them; with no patch built the report
   simply has no 3D section. */
function saveReport() {
  if (!metCnt.auto) { toast('Run the segmentation first'); return; }
  const payload = { patch3d: patch3dParams() };
  beJson('saveReport', JSON.stringify(payload), function (r) {
    if (!r) { toast('⚠ Report export failed', 3800); return; }
    if (r.pending) { showLd('Building the PDF…'); setSt('Building the PDF…', 'busy'); return; }
    if (!r.ok && r.msg === 'Cancelled') return;        // he changed his mind, not an error
    toast((r.ok ? '✓ ' : '⚠ ') + r.msg, r.ok ? 3400 : 3800);
    if (r.ok) { resSaved = true; paintTabs(); }
  });
}

/* The PDF went the long way round, through Chromium, and lands here. */
function onReportDone(ok, msg) {
  hideLd();
  setSt(ok ? 'Report saved ✓' : 'Error', ok ? 'ok' : 'err');
  toast((ok ? '✓ ' : '⚠ ') + msg, ok ? 3400 : 4200);
  if (ok) { resSaved = true; paintTabs(); }
}

function saveGT() {
  beJson('saveGTMask', undefined, function (r) {
    if (!r) { toast('⚠ Could not save the mask', 3800); return; }
    toast((r.ok ? '✓ ' : '⚠ ') + r.msg, r.ok ? 3200 : 3800);
  });
}
